File: 과제7_gLCS/gLCS.py
def fileopen_getstring():
    inp = open("sampleData/1.inp")
    s1, s2 = "".join(input().split()), "".join(input().split())
    inp.close()
    return s1, s2


def fileopen_putstring(answer):
    out = open("gLCS.out", "w")
    #out.write("%s" % answer)
    print(answer)
    out.close()
    return


def LCS():

    def maketable():
        return [[0 for _ in range(len(string2) + 2)] for _ in range(len(string1) + 2)]

    def fill_table(a, b):
        return max(table[a - 1][b - 1], table[a - 2][b - 2], table[a - 1][b - 2], table[a - 2][b - 1]) + 1 \
                if string1[a - 2] == string2[b - 2] else 0

    def makeGLCS(a, b):
        if a < 2 or b < 2 or table[a][b] == 0: return ""
        temp_list = []
        for i in range(1, 3):
            for j in range(1, 3):
                if string1[a - 2] == string2[b - 2] and table[a][b] == table[a - i][b - j] + 1: temp_list.append(makeGLCS(a - i, b - j))
        return min(temp_list) + string1[a - 2] if temp_list else None

    table, string_list, tup, num = maketable(), [], [], 0
    for i in range(2, len(string1) + 2):
        for j in range(2, len(string2) + 2):
            table[i][j] = fill_table(i, j)
            if num < table[i][j]: num, tup = table[i][j], []
            if num == table[i][j]: tup.append((i, j))

    for i in range(len(tup)): string_list.append(makeGLCS(tup[i][0], tup[i][1]))
    return min(string_list) if string_list else None


if __name__ == "__main__":
    string1, string2 = fileopen_getstring()
    fileopen_putstring(LCS())

File: 과제7_gLCS/gLCS.cpp
#include <bits/stdc++.h>

using namespace std;

string A, B;
int alen, blen;
ofstream output;
string SearchStr(vector<vector<int > >& mat, int x, int y) {
    if (x < 2 || y < 2 || !(mat[x][y]))
        return "";
    vector<string> lst;
    string s = "";
    for (int i = 1; i < 3; ++i)
        for (int j = 1; j < 3; ++j)
            if (A[x - 2] == B[y - 2])
                if (mat[x][y] == (mat[x - i][y - j] + 1))
                    lst.push_back(SearchStr(mat, x - i, y - j));
    if (!lst.empty())
        s = lst[0];
    for (int i = 1; i < lst.size(); ++i)
        if (s > lst[i])
            s = lst[i];
    return s + A[x - 2];
}

string LCS() {
    vector<vector<int > > mat(alen + 2, vector<int >(blen + 2, 0));
    vector<string> lst;
    vector<int> x, y;
    string s;
    int mx = 0;
    for (int i = 2; i < alen + 2; ++i)
        for (int j = 2; j < blen + 2; ++j) {
            mat[i][j] = (A[i - 2] == B[j - 2]) * (max(max(mat[i - 1][j - 1], mat[i - 2][j - 2]), max(mat[i - 1][j - 2], mat[i - 2][j - 1])) + 1);
            if (mx < mat[i][j]) {
                mx = mat[i][j];
                x.clear(); y.clear();
            }
            if (mat[i][j] == mx) {
                x.push_back(i); y.push_back(j);
            }
        }

    for (int i = 0; i < x.size(); ++i)
        lst.push_back(SearchStr(mat, x[i], y[i]));
    s = lst[0];
    for (int i = 1; i < lst.size(); ++i)
        if (s < lst[i])
            s = lst[i];
    return s;
}

int main(void) {
    ifstream  input("sampleData/2.inp");
    output.open("1.out");
    string buf;
    input >> A >> B;
    alen = A.length();
    blen = B.length();
    buf = LCS();
    cout << buf << endl;
    output << buf;
    return 0;
}
